Added unstructured linear regridding

schlunma opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

-->

## Description

This PR adds a builtin scheme for unstructured linear regridding, which can be used in the recipe with `scheme: linear`. This scheme allows lazy regridding and weights caching. It uses the super fast Delaunay tesselation of scipy (see [`scipy.spatial.Delaunay`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.Delaunay.html)).

Unlike `esmf_regrid.schemes.ESMFBilinear`, it does not require the data to be UGRID compliant, which makes it much more flexible. For example, it can be used to regrid native ERA5 data in GRIB format (that is on a reduced Gaussian grid), see https://github.com/ESMValGroup/ESMValCore/pull/2178.

Here is a short recipe to test this:

```yaml
# ESMValTool
---
documentation:
  title: Test
  description: Test
  authors:
    - schlund_manuel

datasets:
  - {project: CMIP6, dataset: ICON-ESM-LR, exp: historical, ensemble: r1i1p1f1, grid: gn, timerange: '1850/2014'}

preprocessors:
  regrid:
    regrid:
      target_grid: 1x1
      scheme: linear

diagnostics:
  test:
    variables:
      tas:  # unmasked input data
        mip: Amon
        preprocessor: regrid
      gpp:  # masked input data
        mip: Lmon
        preprocessor: regrid
    scripts:
      null
```

<!--
    Please describe your changes here, especially focusing on why this pull
    request makes ESMValCore better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html

    Please fill in the GitHub issue that is closed by this pull request,
    e.g. Closes #1903
-->

Related to https://github.com/ESMValGroup/ESMValCore/pull/2178

Link to documentation: https://esmvaltool--2350.org.readthedocs.build/projects/ESMValCore/en/2350/recipe/preprocessor.html#built-in-regridding-schemes

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] ~All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful~ Remaining Codacy issue cannot be fixed

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

bouweandela commented on esmvalcore/preprocessor/_regrid_unstructured.py:
```diff
@@ -47,3 +59,309 @@ def regridder(
+            regridded_data = self._regrid_lazy(
+                src_data, udim, self.weights.dtype
+            )
+            # TODO: add limit='128 MiB' to reshape once dask bug is solved
```
I would recommend respecting the configured chunk-size instead of hardcoding 128 MiB

schlunma replied:
Good point. The [doc of reshape](https://docs.dask.org/en/stable/generated/dask.array.reshape.html) suggests that this is done for `limit=None`. However, this may raise the following warning:

```bash
PerformanceWarning: Reshaping is producing a large chunk. To accept the large
chunk and silence this warning, set the option
    >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):
    ...     array.reshape(shape)

To avoid creating the large chunks, set the option
    >>> with dask.config.set(**{'array.slicing.split_large_chunks': True}):
    ...     array.reshape(shape)Explictly passing ``limit`` to ``reshape`` will also silence this warning
    >>> array.reshape(shape, limit='128 MiB')
  return function(items, **kwargs)
```

bouweandela replied:
Would it be possible to do the reshape as part of the regridding instead of afterwards, to avoid this issue? Maybe in the `_interpolate` function?

schlunma replied:
Great idea! Implemented that in https://github.com/ESMValGroup/ESMValCore/pull/2350/commits/fa52c672062c8cfb74b3da9d5d227ef76632bb9c

bouweandela commented on esmvalcore/preprocessor/_regrid_unstructured.py:
```diff
@@ -47,3 +59,309 @@ def regridder(
+
+        # Ensure correct dtype
+        if regridded_data.dtype != cube.dtype:
+            regridded_data = regridded_data.astype(cube.dtype)
```
Is this desirable if the input cube has integer input data?

schlunma replied:
Very good point! No, it's certainly not desired. It would be nice to use the decorator introduced in https://github.com/ESMValGroup/ESMValCore/pull/2393 here, so maybe we can merge this one first?

bouweandela reviewed (changes requested):
Looks good! Two more suggestions for improvement:
1) Could you add some explanation of what regular/irregular/unstructured grids are somewhere?
2) Is there any way for users to see which regridding scheme is used when they call esmvaltool.preprocessor.regrid? If not, a debug log message might be nice.

schlunma commented:
>     2. Is there any way for users to see which regridding scheme is used when they call esmvaltool.preprocessor.regrid? If not, a debug log message might be nice.

There is a debug message [here](https://github.com/ESMValGroup/ESMValCore/blob/933d7ef694af8db44c5a307c1f581bf7470e8491/esmvalcore/preprocessor/_regrid.py#L582), does that work for you?

schlunma commented:
> 1. Could you add some explanation of what regular/irregular/unstructured grids are somewhere?

I just did a quick online search on this to get some commonly accepted definitions for that. Unfortunately, the definitions I found are vastly different to what we are using. [For example](https://en.wikipedia.org/wiki/Regular_grid), in most sources, irregular an unstructured grids are the same.

How should we proceed with this? Use "our" definitions of it or try to use the "official" ones? Please keep in mind that "our" definitions are mentioned all over the place in the documentation...

bouweandela commented:
It would be a great start if you could just describe our conventions. A nice addition would be to explain how they relate to how other software projects describe their grids, e.g https://earthsystemmodeling.org/regrid/#grid-types, https://xesmf.readthedocs.io/en/stable/notebooks/Rectilinear_grid.html#Input-grid, https://xesmf.readthedocs.io/en/stable/notebooks/Curvilinear_grid.html#Input-grid, https://xesmf.readthedocs.io/en/stable/other_tools.html

schlunma commented:
> It would be a great start if you could just describe our conventions. A nice addition would be to explain how they relate to how other software projects describe their grids, e.g https://earthsystemmodeling.org/regrid/#grid-types, https://xesmf.readthedocs.io/en/stable/notebooks/Rectilinear_grid.html#Input-grid, https://xesmf.readthedocs.io/en/stable/notebooks/Curvilinear_grid.html#Input-grid, https://xesmf.readthedocs.io/en/stable/other_tools.html

Done in https://github.com/ESMValGroup/ESMValCore/pull/2350/commits/084d15d49619d714dfa6a74eb20329d090e9f61a

axel-lauer commented:
I just tested the new regridder and it works nicely. The values and distribution look fine but I was a bit puzzled about the missing top right and bottom right edges, e.g. when regridding tas which does (should) not have any missing values. This looks a bit strange and might need to be double checked, please see this example here:

![ICON-tas-regrid](https://github.com/ESMValGroup/ESMValCore/assets/25367775/7a445ba7-e8d6-4648-a4f1-aaca3d172939)

schlunma commented:
Thanks @axel-lauer for reviewing! I fixed this issue in https://github.com/ESMValGroup/ESMValCore/pull/2350/commits/e14e3b5f537324488d6347acbfdd4b034dfc36ad. Below you can see a regridded ERA5 GRIB file with the new version of the code:

![grafik](https://github.com/ESMValGroup/ESMValCore/assets/32543114/d83b686e-8d7e-453d-81b6-3632bc195224)

The problem was that the old code did not consider the periodic boundary conditions on the sphere. I fixed this by adding the convex hull of the source points (i.e., the outer edge) of all points shifted by +/-360° to the list of source points. This might not be the most elegant solution, but it does the job.

schlunma commented:
@valeriupredoi would you do us the honor of merging this? 🍻

valeriupredoi commented:
done! Hope that curvy corner got fixed :laughing:
